PA193_mnemonic_Slytherin/mnemonic.py

```diff
@@ -26,15 +25,87 @@
 SEED_LEN = 64
 
 
-def _generate_seed(mnemonic: str, seed_password: str = '') -> bytes:
+class Mnemonic:
+    """Class for mnemonic representation.
+    """
+    
+    def __init__(self, mnemonic: str):
+        """Convert mnemonic phrase to entropy using dictionary to ensure its validity.
+        :raises ValueError: on invalid parameters
+        """
+        if not isinstance(mnemonic, str):
+            raise ValueError('Cannot instantiate mnemonic')
+    
+        words = mnemonic.split()
+        l = len(words)
+        if not l in (12, 15, 18, 21, 24):
+            raise ValueError('Cannot instantiate mnemonic')
+
+        _, word_dict = __get_dictionary()
+        try:
+            indexes = [word_dict[word] for word in words]
+        except KeyError:
+            raise ValueError('Cannot instantiate mnemonic')
+
+        # Concatenate indexes into single 
+        indexes_bin = sum([indexes[-i - 1] << i * 11 for i in reversed(range(l))])
+
+        # Number of bits entropy is shifted by
+        shift = l * 11 // 32
+
+        checksum = indexes_bin & (pow(2, shift) - 1)
+        entropy_bin =  indexes_bin >> shift
+        entropy = entropy_bin.to_bytes((l * 11 - shift) // 8, byteorder='big')
+
+        # Check correctness
+        check = __get_checksum(entropy, shift)
+        if check != checksum:
+            raise ValueError('Cannot instantiate mnemonic')
+    
+        self._entropy = Entropy(entropy)
+        self._str = mnemonic
```

Comment:
If I understant this correctly as we talked about this earlier... In order to validate mnemonic phrase, it is requred to generate entropy from the mnemonic and then checksum from entropy? So you need entropy to validate mnemonic?

Reply:
I wonder whether Entropy and Mnemonic aren't actualy the same core thing with just diferent representation. Mnemonic is 1 to 1 with Entropy, right? Semantic of mnemonic is that it's just more convenient representation of entropy. What if we would have just one class `Entropy(bytes)`? Then it would have the code above (currently `Mnemonic.__init__`) as `Entropy.from_mnemonic(cls, mnemonic: str)` which would instantiate `Entropy` from given `mnemonic`.

`generate(entropy: bytes, seed_password: str = '')` would then instantiate entropy based on provided bytes, and then call `entropy.mnemonic` which would be read-only property to get mnemonic from internal representation of entropy. This cold have cache so it's always computer only once.

`recover(mnemonic: str, seed_password: str = '')` would call `Entropy.from_mnemonic(cls, mnemonic: str)` which then saves us from `entropy = __mnemonic2entropy(mnemonic)`.

`_generate_seed(entropy., seed_password)` would do
~~~py
pbkdf2_hmac('sha512', entropy.mnemonic,encode(), ("mnemonic" + seed_password),encode(), PBKDF2_ROUNDS, SEED_LEN)
~~~
But `_generate_seed(entropy., seed_password)` would be incorporated to e. g. `Seed.from_entropy` factory method or `Entropy.seed` read-only property with cache.

`verify(mnemonic: str, expected_seed: bytes, seed_password: str = '')` would call `Entropy.from_mnemonic(cls, mnemonic: str)`

What do you think? Hot or not? 😄

Reply:
I think that, although its true they are just different representations, it makes it cleaner and easier to read. We also have the nice clean api where everything has the type you'd expect.

Reply:
Its deffinitely possible to do, but i would rather make them inherit common base class which would resolve my "problem" with dictionary just the same.